sysusers: allow uid:gid in sysusers.conf files

mvo5 opened the pull request:
This PR allows to write sysuser.conf lines like:
```
u games 5:60 -
```
This will create an a "games" user with uid 5 and games group with
gid 60. This is arguable ugly, however it is required to represent
certain configurations like the default passwd file on Debian and
Ubuntu.

When the ":" syntax is used and there is a group with the given
gid already then no new group is created. This allows writing the
following:
```
g unrelated 60
u games 5:60 -
```
which will create a "games" user with the uid 5 and the primary
gid 60. No group games is created here (might be useful for [1]).

[1] https://pagure.io/packaging-committee/issue/442

mvo5 commented:
The relevant issue is https://github.com/systemd/systemd/issues/7967 - I would love to write a integration test, however I am not familiar with the prefered way that systemd is using. Any hints/pointers are appreciated.

poettering commented on src/sysusers/sysusers.c:
```diff
@@ -1550,6 +1553,17 @@ static int parse_line(const char *fname, unsigned line, const char *buffer) {
+                                if (split_pair(resolved_id, ":", &uid, &gid) == 0) {
+                                        r = parse_gid(gid, &i->gid);
+                                        if (r < 0) {
+                                                log_error("Failed to parse GID: %s", id);
```
(and yeah, I am aware the UID parsing code doesn't do this... but it really should, and for new code we should be stricter on this...)

(and of course you get extra points if you fix the UID parsing accordingly, too.. ;-)

mvo5 replied:
Good point, thank you. I updated the error message (and the one below). I had a quick look in the file but it looks no other log message needs updating.

poettering commented on src/sysusers/sysusers.c:
```diff
@@ -1550,6 +1553,17 @@ static int parse_line(const char *fname, unsigned line, const char *buffer) {
+                                        }
+                                        i->gid_set = true;
+                                        i->gid_existing_ok = true;
+                                        resolved_id = uid;
```
this will result in a double free like this, as well as leaked memory...

You need to free the old resolved_id first, before you override its pointer. And you need to assign NULL to uid, after you stole its object. Note that you are using the gcc cleanup logic here, i.e. the strings are marked _cleanup_free_ and thus will be destroyed when you leave the scope. And this means the data pointed to initially by resolved_id will be leaked as it is never freed, and the data assigned to it later will be freed twice: once in the inner block as "uid", and once in the outer block again...

mvo5 replied:
Thanks for your careful review and sorry for so many errors in so few lines :/ (too much work in languages with automatic memory management recently). I fixed them as you suggested.  I appreciate a careful re-review.

My original version had more duplication but less leaks/double-frees:
```                               
                               _cleanup_free_ char *uid = NULL, *gid = NULL;
                               if (split_pair(resolved_id, ":", &uid, &gid) == 0) {
                                        r = parse_uid(uid, &i->uid);
                                        if (r < 0) {
                                                log_error_errno(r, "Failed to parse UID: '%s': %m", id);
                                                return -EBADMSG;
                                        }
                                        i->uid_set = true;

                                        r = parse_gid(gid, &i->gid);
                                        if (r < 0) {
                                                log_error_errno(r, "Failed to parse GID: '%s': %m", id);
                                                return -EBADMSG;
                                        }
                                        i->gid_set = true;
                                        i->gid_existing_ok = true;
                                } else {
                                        r = parse_uid(resolved_id, &i->uid);
                                        if (r < 0) {
                                                log_error_errno(r, "Failed to parse UID: '%s': %m", id);
                                                return -EBADMSG;
                                        }
                                        i->uid_set = true;
                                }
```
But I guess the current (fixed) code is better as it avoids those duplications.

mvo5 commented on man/sysusers.d.xml:
```diff
@@ -189,7 +189,10 @@ u      root   0               "Superuser"           /root</programlisting>
       path's owner/group. This is useful to create users whose UID/GID
       match the owners of pre-existing files (such as SUID or SGID
-      binaries).</para>
+      binaries). The syntax "uid:gid" is also supported to allow
```
Indeed, thank you. I fixed this now.

yuwata commented on src/sysusers/sysusers.c:
```diff
@@ -1550,9 +1553,22 @@ static int parse_line(const char *fname, unsigned line, const char *buffer) {
+                                        i->gid_existing_ok = true;
+                                        mfree(resolved_id);
+                                        resolved_id = uid;
+                                        uid = NULL;
```
Please use `free_and_replace(resolved_id, uid);` instead of the above three lines.

mvo5 replied:
Thanks, fixed now. Nice helper, I was not aware of it.

mvo5 commented on man/sysusers.d.xml:
```diff
@@ -189,7 +189,10 @@ u      root   0               "Superuser"           /root</programlisting>
       match the owners of pre-existing files (such as SUID or SGID
-      binaries).</para>
+      binaries). The syntax "uid:gid" is also supported to allow
+      creating a gid with a different ID from uid. If the gid in
```
Thanks for your suggestions, I updated the text.

yuwata commented on src/sysusers/sysusers.c:
```diff
@@ -1550,9 +1553,22 @@ static int parse_line(const char *fname, unsigned line, const char *buffer) {
+                                        r = parse_gid(gid, &i->gid);
+                                        if (r < 0) {
+                                                log_error_errno(r, "Failed to parse GID: '%s': %m", id);
+                                                return -EBADMSG;
```
I think it is not necessary to return `-EBADMSG` here. Please use `return log_error_errno(...);`.

mvo5 replied:
Thank you! Cleaner indeed.

yuwata commented on src/sysusers/sysusers.c:
```diff
@@ -1550,9 +1553,22 @@ static int parse_line(const char *fname, unsigned line, const char *buffer) {
                                 if (r < 0) {
-                                        log_error("Failed to parse UID: %s", id);
+                                        log_error_errno(r, "Failed to parse UID: '%s': %m", id);
                                         return -EBADMSG;
```
Same as the above.

yuwata commented on src/sysusers/sysusers.c:
```diff
@@ -1602,7 +1618,7 @@ static int parse_line(const char *fname, unsigned line, const char *buffer) {
                                 if (r < 0) {
-                                        log_error("Failed to parse GID: %s", id);
+                                        log_error_errno(r, "Failed to parse GID: '%s': %m", id);
                                         return -EBADMSG;
```
Same as the above.

yuwata commented on src/sysusers/sysusers.c:
```diff
@@ -1550,10 +1553,19 @@ static int parse_line(const char *fname, unsigned line, const char *buffer) {
+                                        r = parse_gid(gid, &i->gid);
+                                        if (r < 0) {
+                                                return log_error_errno(r, "Failed to parse GID: '%s': %m", id);
+                                        }
```
When only single line in `if()` block, then please do not use `{}`. Of course this is valid. But please read our CODING_STYLE.

mvo5 replied:
Sure, fixed! Sorry for this.

yuwata commented on src/sysusers/sysusers.c:
```diff
@@ -1550,10 +1553,19 @@ static int parse_line(const char *fname, unsigned line, const char *buffer) {
-                                        log_error("Failed to parse UID: %s", id);
-                                        return -EBADMSG;
+                                        return log_error_errno(r, "Failed to parse UID: '%s': %m", id);
                                 }
```
Same here.

yuwata commented on src/sysusers/sysusers.c:
```diff
@@ -1602,8 +1614,7 @@ static int parse_line(const char *fname, unsigned line, const char *buffer) {
-                                        log_error("Failed to parse GID: %s", id);
-                                        return -EBADMSG;
+                                        return log_error_errno(r, "Failed to parse GID: '%s': %m", id);
                                 }
```
Same here.

yuwata commented:
Thanks. Now, the code base itself looks good to me. I will test this.

poettering commented on man/sysusers.d.xml:
```diff
@@ -189,7 +189,10 @@ u      root   0               "Superuser"           /root</programlisting>
-      binaries).</para>
+      binaries).
+      The syntax <literal><replaceable>uid</replaceable>:<replaceable>gid</replaceable></literal> is also supported to
+      allow creating user and group pairs with different numeric UID and GID values. If a group with the indicate GID already exists it is not created.
```
indicate → indicated

mvo5 commented:
@yuwata Thank you. If I can help in any way with the test (or if there is a standard way to write integration tests) please do let me know. Happy to look into this!

poettering commented:
> @yuwata Thank you. If I can help in any way with the test (or if there is a standard way to write integration tests) please do let me know. Happy to look into this!

well, not sure about "standard", but there's the stuff in test/TEST-XY-*. Would be excellent if you would add a test there! see test/README.testsuite for a brief /and sightly out-of-date) intro

yuwata commented:
Hmm,
```
g hoge    300     -            -
u foo     301     -            -

g aaa     302     -            -
u aaa     303:302 -            -
u bbb     304:302 -            -
u ccc     305:306 -            -
```
Then I get
```
$ sudo SYSTEMD_LOG_LEVEL=debug ./systemd-sysusers /tmp/test-7973.conf 
Creating group hoge with gid 300.
Creating group foo with gid 301.
Creating user foo (n/a) with uid 301 and gid 301.
Creating group bbb with gid 302.
Creating user bbb (n/a) with uid 304 and gid 302.
Suggested group ID 302 for aaa already used.
Creating user aaa (n/a) with uid 303 and gid 302.
Creating group ccc with gid 306.
Creating user ccc (n/a) with uid 305 and gid 306.
```
I thought that group aaa should be created instead of bbb.

yuwata commented:
```
$ cat /tmp/test-7973.conf 
g hoge    300     -            -
u foo     301     -            -

g baz     302     -            -
u aaa     303:302 -            -
u bbb     304:302 -            -
u ccc     305:306 -            -
```
```
$ sudo SYSTEMD_LOG_LEVEL=debug ./systemd-sysusers /tmp/test-7973.conf 
Creating group baz with gid 302.
Creating group hoge with gid 300.
Creating group ccc with gid 306.
Creating user ccc (n/a) with uid 305 and gid 306.
Suggested group ID 302 for aaa already used.
Creating user aaa (n/a) with uid 303 and gid 302.
Creating group foo with gid 301.
Creating user foo (n/a) with uid 301 and gid 301.
Suggested group ID 302 for bbb already used.
Creating user bbb (n/a) with uid 304 and gid 302.
```
Is it expected?

yuwata commented:
Before this PR, each 'u' line always create both user and group, so skipping to process 'g' line with the same name in `process_item()` may be a good idea. However, with this PR, now 'u' line may reuse a already existing group and may not create group with the same name. So, I think `g` line needs always to actually create the group. Moreover, as my previous tests, now multiple 'u' line use the same group. So, at least when gid is specified, the corresponding group should not created by 'u' line. In other words. when 'uid:gid' format is used, then we should require that the group is explicitly specified in 'g' line or is already exists. Otherwise, as the previous test results, it is almost not possible to predict the created group name of the gid. What do you think?

yuwata commented:
So, my suggesting change is the following.
- In `process_item()`, `ADD_GROUP` always call `add_group()`.
- In `add_user()`, at first, check the existence of the group whose gid is `i->gid`.

yuwata commented:
Excellent! Please rebase this on recent snapshot.

yuwata commented:
Hmm. Using ordered hashmap does not cover my concerns.
Please add a test something like below.
```
g xxx 310
u yyy 311:310
u xxx 312:310
```
I have not tested the above. But I guess the results are different from the expectation.

mvo5 commented:
@yuwata Thank you! I will add more tests and fix things. Sorry for the not-rebased push earlier.

mvo5 commented:
Fwiw, happy to squash my commits into a single one, I mostly did this sequence to make the review easier. I added a test based on the comment in https://github.com/systemd/systemd/pull/7973#issuecomment-360089481 and did "In process_item(), ADD_GROUP always call add_group()" - I'm not sure we need "In add_user(), at first, check the existence of the group whose gid is i->gid." anymore now that the processing is ordered. But I'm happy to dig into this if I'm wrong.

poettering commented:
we generally prefer reviewing/merging "perfect" patch sets, i.e.where each individual patch that makes up a PR makes logical sense on its own, but does not reflect historical development progress. i.e. we don't want to see commits that correct earlier broken commits in the same PR. Or in other words, it's nice to split up PRs into multiple commits, but please make sure that everything remains perfectly bisectable

yuwata commented:
I've not tested, but I guess that the following two examples make different results.
```
u aaa 300:350
u bbb 301:350
```
and 
```
u bbb 301:350
u aaa 300:350
```
So, I think we should not implicitly create group by 'u' line with 'uid:gid' format.

mvo5 commented:
@yuwata Thank you! Indeed, the two lines you posted will have different results with the previous iteration of this PR (predictable thanks to the ordering but still different). I agree this is confusing and I now added code to ensure the gid exists and also added a corresponding test (and one more negative testcase to exercise this mechanism).

poettering commented on src/sysusers/sysusers.c:
```diff
@@ -1099,10 +1099,17 @@ static int add_group(Item *i) {
                         return log_error_errno(r, "Failed to verify gid " GID_FMT ": %m", i->gid);
+                /* If we require the gid to already exist we can return here (either with an error or with nothing to do) */
+                if (i->gid_must_exist) {
+                        if (r > 0) {
```
hmm, isn't this check reversed? not following here?

mvo5 replied:
Sorry, I'm not sure I understand. I improved the code comment but maybe I'm missing something. Happy to rework if you think this is confusing.

poettering replied:
ah, i was a bit confused. thanks for the longer comment, that helped me grok this.

poettering commented:
lgtm, @yuwata is this to your liking too? If so, let's get this merged and we can still include it in v237 (which we hopefully can release this WE)

yuwata commented:
Yeah, looks good to me too! Let's merge this. Thanks!
